Reworking how integrity check is done.

adiom-mark opened the pull request:
This is sort of a rework to simplify integrity check implementation on the connector side by making the integrity check call from the coordinator synchronous. So the core change is that a connector must simply answer the integration check for a single read plan task. The coordinator is responsible for resolving these and doing any parallelization of tasks.

Currently the integrity check is simplified to do just the count, and the read plan is based off the source. The check is not quite the same as before, as extra items in the destination will not be detected. To get similar behavior, we'd need to run an integrity check with the dst and src swapped.

Next step would be to build on this and add ability to do a more comprehensive integrity check (e.g. hash of the data)

alex-thc commented on coordinators/simple/coordinator.go:
```diff
@@ -25,11 +28,17 @@ type Simple struct {
 }
 
-func NewSimpleCoordinator() *Simple {
+func NewSimpleCoordinator(numParallelIntegrityCheckTasks int) *Simple {
```
this feels like it should be a CoordinatorSettings struct rather than an int param

adiom-mark replied:
we don't have one right now, so didn't think it needed to be there yet. i think this is an easy refactor later

adiom-mark commented:
Updated so that the integration check is cancellable, and uses a context for it. The context for this is separate, so for now we don't mess with how shutdown works for other things. When canceling a check, it will return the result so far.

Updated documentation, and various other minor stuff.

As each task completes we now display some text so that can serve as a sort of progress marker.

Will defer the count optimizations to a future pr.

alex-thc commented on coordinators/simple/coordinator.go:
```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
+				}
+				matches := bytes.Equal(srcRes.Digest, dstRes.Digest) && srcRes.Count == dstRes.Count
+				if !matches {
+					slog.Info(fmt.Sprintf("Mismatch %v, src: %v, dest: %v", task, srcRes, dstRes))
```
I'd make it WARN

alex-thc commented on coordinators/simple/coordinator.go:
```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
+					slog.Info(fmt.Sprintf("Mismatch %v, src: %v, dest: %v", task, srcRes, dstRes))
+					return mismatchErr
+				}
+				slog.Info(fmt.Sprintf("Matched %v", task))
```
This probably should be DEBUG

adiom-mark replied:
this is essentially the poor man's progress report

alex-thc commented on coordinators/simple/coordinator.go:
```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
+						slog.Info(fmt.Sprintf("Source integrity task canceled %v", task))
+						return nil
+					}
+					slog.Error(fmt.Sprintf("Source integrity check failed: %v", err))
```
Would be useful to know for which task

adiom-mark replied:
agreed, that'd be useful, but probably will add in a future pr

alex-thc commented on coordinators/simple/coordinator.go:
```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
+						slog.Info(fmt.Sprintf("Destination integrity task canceled %v", task))
+						return nil
+					}
+					slog.Error(fmt.Sprintf("Destination integrity check failed: %v", err))
```
Would be useful to know for which task

adiom-mark commented on coordinators/simple/coordinator.go:
```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
+				srcRes, err := src.Endpoint.IntegrityCheck(ctx, fid, task)
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
+						slog.Info(fmt.Sprintf("Source integrity task canceled %v", task))
```
i made this an info because if you cancel it, you may want to know where things got canceled.
